File: assignment_6/sentimental-credit/credit.py
from cs50 import get_string
import re


def Luhn(number):
    sum = 0
    size = len(number)
    for i in range(size):
        if i % 2 != 0:
            if (int(number[size-i-1])*2 < 10):
                sum += 2*int(number[size-i-1])
            else:
                sum += (1+((int(number[size-i-1])*2)-10))
        else:
            sum += int(number[size-i-1])
    if sum % 10 == 0:
        return True
    else:
        return False


card = ""

number = get_string("Number: ")

if re.findall("^[3][4,7]", number) and len(number) == 15 and Luhn(number) == True:
    card = "AMEX"

elif re.findall("^[5][1-5]", number) and len(number) == 16 and Luhn(number) == True:
    card = "MASTERCARD"

elif re.findall("^4", number) and (len(number) == 13 or len(number) == 16) and Luhn(number) == True:
    card = "VISA"

else:
    card = "INVALID"

print(card)

File: assignment_1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main(void){
    long card_number = get_long("Card number:");

    int number_of_digits=log10(card_number)+1;

    int first_digits=card_number%((long) pow(10,number_of_digits))/ (long) pow(10,number_of_digits-2);

    //printf("number of digits %i\n",number_of_digits);
    //printf("first digits %i\n",first_digits);

    //check for card type
    string card_type;

    if(number_of_digits==15 && (first_digits==34 || first_digits==37)){
        card_type="AMEX";
    }

    else if(number_of_digits==16 && (first_digits>=51 && first_digits<=55)){
        card_type="MASTERCARD";
    }

    else if((number_of_digits==13 || number_of_digits==16) && (first_digits>=40 && first_digits<=49)){
        card_type="VISA";
    }

    else{
        card_type="INVALID";
        printf("%s\n",card_type);
        return 0;
    }


    //Luhn's Algorithm
    int sum=0;
    //Start in 2nd to last digit
    for(int i=0; i<number_of_digits; i++){
        int digit=card_number%((long) pow(10,(i+1)))/ (long) pow(10,i);
        if (i==0||i%2==0){
            sum=sum+digit;
        }
        else if(i%2!=0){
            if(digit*2<10){
                sum=sum+digit*2;
            }
            else{
                sum=sum+(digit*2)%10+(digit*2)%100/10;
            }
        }
    }
    //printf("Value of sum: %i\n",sum);
    if(sum%10!=0){
        card_type="INVALID";
    }

    printf("%s\n",card_type);
}
